src/storage/src/upsert_continual_feedback.rs

```diff
@@ -0,0 +1,826 @@
+                                }
+                            }
+                            AsyncEvent::Progress(upper) => {
+                                tracing::trace!(?upper, "received progress in upsert");
+
+                                // Ignore progress updates before the `resume_upper`, which is our initial
+                                // capability post-snapshotting.
+                                if PartialOrder::less_than(&upper, &resume_upper) {
+                                    tracing::trace!(?upper, ?resume_upper, "ignoring progress updates before resume_upper");
+                                    continue;
+                                }
+
+                                if let Some(ts) = upper.as_option() {
+                                    tracing::trace!(?ts, "downgrading output capability");
+                                    let _ = output_cap.try_downgrade(ts);
+                                }
+                                input_upper = upper;
+                            }
+                        }
+                        let events_processed = i + 1;
+                        if let Some(max) = snapshot_buffering_max {
+                            if events_processed >= max {
+                                break;
+                            }
+                        }
+                    }
+                }
+            };
+
+            // We try and drain from our stash every time we go through the
+            // loop. More of our stash can become eligible for draining both
+            // when the source-input frontier advances or when the persist
+            // frontier advances.
+
+            drain_staged_input::<_, G, _, _, _>(
+                &mut stash,
+                &mut commands_state,
+                &mut output_updates,
+                &mut multi_get_scratch,
+                DrainStyle::ToUpper{input_upper: &input_upper, persist_upper: &persist_upper},
+                &mut error_emitter,
+                &mut state,
+            )
+            .await;
+
+            tracing::trace!(?output_updates, "output updates for complete timestamp");
+
+            // TODO: Feels somewhat inefficient to be peeling
+            // of and creating new vecs per group of
+            // timestamp. Especially since at steady state
+            // we're expecting to only have one group...
+            while !output_updates.is_empty() {
+                let cap = output_updates[0].1.clone();
+                let ts = cap.time();
+
+                let partition_point = output_updates
+                    .partition_point(|update| update.1.time() == ts);
+
+                let mut ts_updates: Vec<_> = output_updates.drain(0..partition_point)
+                    .map(|(update, cap, diff)| (update, cap.time().clone(), diff))
+                    .collect();
+
+                output_handle.give_container(&cap,&mut ts_updates);
```

Comment:
Using `give_container` has the risk of creating too big containers. Use the normal `.give()` API here so that timely does its chunking automatically.